### PR TITLE
[inductor] correctly infer dtype of `full`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #95593

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Reported by Natalia here: https://github.com/pytorch/pytorch/pull/95176#issuecomment-1441094262

File: test/test_dynamic_shapes.py [hunk 1]
```diff
@@ -801,5 +803,47 @@ def is_complex(x):
                 self.assertEqual(op.is_integer, None)
                 self.assertTrue(op.is_real)
 
+class TestFull(TestCase):
```
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:806]: Need to move this to test_torchinductor/_dynamic_shapes

**Review by ezyang — APPROVED** [review]
seems fine

**Review by ngimel — APPROVED** [review]

**Review by lezcano — COMMENTED** [review]
Using the PrimTorch version should fix this, and I bet it generates the same code.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -1900,8 +1900,27 @@ def copy_strided(x, stride):
     return ir.ExternKernel.require_stride_order(x, stride_order)
 
 
+# Performs dtype inference for full
+def infer_full_dtype(fill_value, dtype):
```
Comment (ngimel) [inline torch/_inductor/lowering.py:1904]: Better to use _prims_common.type_to_dtype instead

**Pull request closed without merging**